### PR TITLE
Monorepo ed/x/curve25519

### PR DESCRIPTION
**Pull request by pinkforest**
As discussed this is the monorepo PR
- This would need to be merged - not squashed - in order to retain the combined history from x/ed/curve
- Renames for each crates in separate commit
- Merge commits for each combine
- CI re-organized for Monorepo structure
- This PR needs to be kept minimal and can follow-up with e.g. proper monorepo root README.md etc.
- This would help us keep the crates consistent and synchronized across and messing with git patching
- I was hoping commits could be prefixed-amended based on crate e.g. `ed25519` could have `ed:` prefix and the PR links would be linked to old place - nonetheless this would be the tradeoff with history we would pay for easier consistency across the crates and usually the original commits has all the information from PR's

git combine Procedure (see near last commits)
```
git mv [curve25519-dalek-stuff] curve25519-dalek/
git commit -m 'Workspace curve25519 under curve25519-dalek'

git remote add ed https://github.com/dalek-cryptography/ed25519-dalek.git
git fetch ed
git merge ed/main --allow-unrelated-histories
git mv [ed25519-dalek-stuff] ed25519-dalek/
git commit -m 'Workspace ed25519 under ed25519-dalek'

git remote add x https://github.com/dalek-cryptography/x25519-dalek.git
git fetch x
git merge x/main --allow-unrelated-histories
git mv [x25519-dalek-stuff] x25519-dalek/
git commit -m 'Workspace x25519 under x25519-dalek'
```

### CODE REVIEW

**Comment by rozbb** [comment]
Thank you! Two things:

1. Is there anything we can do to merge the issues/Pars (current and past) from all the constituent repos? Or are we just stuck keeping or historical record in an archived repo?
2. If it's not too annoying, I'm gonna do the git merge steps myself. Just bc it's literally impossible to audit on my end. I'll cherry pick the follow-up commits you have here.

**Comment by pinkforest** [comment]
1. No, that would be modifying the history and that would be the tradeoff as flagged.
2. I have other changes too incl. CI and you would have to audit in any case even if it's your own work :)

You can audit it with rsync:

$ rsync -rcnv a/* b/
```
-r will recurse into the directories
-c will compare based on file checksum
-n will run it as a "dry run" and make no changes, but just print out the files 
   that would be updated
-v will print the output to stdout verbosely
```

e.g. like this:
```bash
git clone https://github.com/dalek-cryptography/curve25519-dalek.git
git clone https://github.com/dalek-cryptography/ed25519-dalek.git
git clone https://github.com/dalek-cryptography/x5519-dalek.git
pull this pr to "monorepo"

rsync -rcnv ../monorepo/curve25519-dalek/* curve25519-dalek/
rsync -rcnv ../monorepo/ed25519-dalek/* ed25519-dalek/
rsync -rcnv ../monorepo/x25519-dalek/* x25519-dalek/
```

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-name: Rust
+name: All
```
Comment (rozbb) [inline .github/workflows/rust.yml:1]: Nit: can we rename this file to something like `workspace.yml` or `common.yml` or something? And a comment somewhere explaining "these tests operate on the entire workspace. they mostly check that everything builds and looks good"

File: README.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# dalek curve25519
+
+TBD
```
Comment (rozbb) [inline README.md:3]: I'll take a crack at writing this

**Review by rozbb — COMMENTED** [review]
Left some comments. Verified that the subdirectories match the current `main` branch of all the relevant repos (outside the necessary Cargo.toml changes). Used
```console
diff -r ./curve25519-dalek ../curve25519-dalek
```
etc.

**Comment by rozbb** [comment]
Alright, made a pretty barebones [README](https://github.com/dalek-cryptography/curve25519-dalek/blob/4a4c288fe6a8fe18a11a2102c326f5e25f8f5597/README.md), but it should be OK for now imo. The dalek image doesn't load correctly anymore bc the URL is gonna change in this PR.

File: .github/workflows/rust.yml [hunk 3]
```diff
@@ -196,35 +60,12 @@ jobs:
...
-    - uses: dtolnay/rust-toolchain@nightly
-    - run: cargo -Z minimal-versions check --no-default-features --features serde
-    # Now check that `cargo build` works with respect to the oldest possible
-    # deps and the stated MSRV
-    - uses: dtolnay/rust-toolchain@1.60.0
-    - run: cargo build --no-default-features --features serde
-    # Also make sure the AVX2 build works
-    - run: cargo build --target x86_64-unknown-linux-gnu
-
-  bench:
```
[the remark this answers is not in the document]
Comment (pinkforest) [reply .github/workflows/rust.yml:215]: Fixed.

**Comment by pinkforest** [comment]
Ok - I had to re-organize commits since this PR needs a merge (not a squash) and we have renames which need to be kept in separate commits in the middle. I gave you co-authored-by since I had to amend the commits in the middle and re-add yours.

**Review by rozbb — APPROVED** [review]
Gotta re-check the subdirs with `diff` again bc of the force push. But barring that good to go @tarcieri ?

Update: re-verfified. For concreteness, the current HEADs are:
* https://github.com/dalek-cryptography/curve25519-dalek/commit/2e3212b8cc325737ae287e94deea55ed26546c58
* https://github.com/dalek-cryptography/ed25519-dalek/commit/58a967f6fb28806a21180c880bbec4fdeb907aef
* https://github.com/dalek-cryptography/x25519-dalek/commit/8613b5a809c78160a2c16b4635f665a15aea2055

**Review by tarcieri — APPROVED** [review]

**Comment by rozbb** [comment]
Build's broken now but I have no idea what's causing it 😕

**Comment by pinkforest** [comment]
Can we please not keep adding commits on top. something broke in transient dependency somewhere for the MSRV build - this should be mergeable in any case as this doesn't cause it.

**Comment by pinkforest** [comment]
It's this one
https://github.com/dtolnay/quote/commit/0c68465f01aed11ce1f007ca1be3e84821224ccb

It will get fixed when quote 0.29 release gets picked up by the transient dep

**Pull request merged**